File: iosApp/iosApp/Services/CultService/CultProposal.swift
// Created by web3d3v on 04/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation

struct CultProposal {
    
    let id: String
    let title: String
    let approved: Double
    let rejeceted: Double
    let endDate: Double // timeIntervalSince1970
    
    let guardianInfo: GuardianInfo?

    let projectSummary: String
    let projectDocuments: [ProjectDocuments]
    
    let cultReward: String
    let rewardDistributions: String
    let wallet: String
    let status: Status
    let stateName: String
    
    struct GuardianInfo {
        
        let proposal: String // aka: name
        let discord: String
        let address: String
    }
        
    struct ProjectDocuments {
        
        let name: String
        let documents: [Document]
        
        enum Document {

            case link(displayName: String, url: URL)
            case note(String)
        }
    }
    
    enum Status {
        case pending
        case closed
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/services/cult/CultProposal.kt
package com.sonsofcrypto.web3walletcore.services.cult

import kotlinx.serialization.Serializable

@Serializable
data class CultProposal(
    val id: String,
    val title: String,
    val approved: Double,
    val rejected: Double,
    val endDate: Double, // timeIntervalSince1970
    val guardianInfo: GuardianInfo?,
    val projectSummary: String,
    val projectDocuments: List<ProjectDocuments>,
    val cultReward: String,
    val rewardDistributions: String,
    val wallet: String,
    val status: Status,
    val stateName: String,
) {
    @Serializable
    data class GuardianInfo(
        val proposal: String, // aka: name
        val discord: String,
        val address: String,
    )

    @Serializable
    data class ProjectDocuments(
        val name : String,
        val documents: List<Document>,
    ) {
        @Serializable
        sealed class Document {
            @Serializable
            data class Link(val displayName: String, val url: String): Document()
            @Serializable
            data class Note(val note: String): Document()
        }
    }

    @Serializable
    enum class Status { PENDING, CLOSED }
}
